### ISSUE TITLE
[Feature] Install apps into `current` folder

### ISSUE THREAD

**Issue by pukkandan** [ScoopInstaller/Scoop#5234]
#### Is your feature request related to a problem? Please describe.

Currently, apps are installed to a versioned folder and a symlink for it is created to `current`. This allows using `current` as a version independent path in most cases. However, some apps (AVs, firewalls and anything that inspects processes; eg **simplewall**) resolve the symlinks internally and only allows pointing to the actual exe path in the versioned folder.

In the case of simplewall, for example, this means that firewall rules have to be manually edited after updating each app

#### Describe the solution you'd like

Install directly into `current` and symlink it to the version folder instead.

Both this problem and this same solution was previously discussed in #2365, but it was then closed without any explanation

#### Describe alternatives you've considered

Couldn't think of any other solution

**Comment by rasa** [comment]
I agree. Every time I upgrade certain apps, I have to update my firewall. Huge pita.

**Comment by niheaven** [comment]
Could you try `scoop config no_junction true`?

**Comment by rasa** [comment]
> Could you try `scoop config no_junction true`? This will trigger `junction` feature and link shim to versioned directory.

@niheaven That doesn't create the `current` directory, it simply links the shim to the specific version. So it doesn't solve the problem.

**Comment by niheaven** [comment]
Okay, I got your idea. The solution needs creating a permanent location so AV apps could remember it once and forever, right?

**Comment by rasa** [comment]
> Okay, I got your idea. The solution needs creating a permanent location so AV apps could remember it once and forever, right?

Yup, but not just AV, but also firewall settings, and maybe even directory/file ACL settings.

Without being critical, I honestly don't understand why we effectively recreate the `current` directory every time we upgrade. IMO, the only reason to do this is if we want to easily return to an older version. But we don't even have a command to do that. And I can't remember even needing to do myself, for any app, scoop-installed or otherwise.

So why don't we simply install into `current`? Or even `~/scoop/app/appname`? Then we don't even need a `persist` folder, as `current` is now "persisted".

Of course, before upgrading to version `y.z`, we could backup the current directory, with:
`cp ~/scoop/app/appname/current/* ~/scoop/app/appname/x.y`
But honestly, if the user needs to downgrade to version `x.y`, I wouldn't restore that `x.y` directory. Instead, I would install version `x.y` from the download cache into `current`. Otherwise, any files the user updated in `current` would get wiped out with the version in `x.y/`.

Of course, if there are files that are created on first install, but shouldn't be overwritten on upgrades/downgrades/installs, we protect them from getting clobbered. I guess we should even persist them if the user uninstalls/reinstalls. To persist these files, we backup the "persisted" files in `current`, allow them to be overwritten on upgrade/downgrade, and then restore them from the backup.

To persist across uninstalls/reinstalls, we simply delete the files the shims link to, and the shims when uninstalling, leaving everything else in `current` alone. On reinstalls, we don't clobber what's already in `current`.

And lastly, scoop needs to know what version is currently installed, so on installs/upgrades/downgrades we remove any prior junctions, and junction the currently installed version (say `~/scoop/app/appname/y.z`) to `current`. If the junction disappears, or `no_junctions` is set, we look to the `.json` file in `current` for the current version.

Or am I missing something basic here?

**Comment by niheaven** [comment]
> But we don't even have a command to do that. And I can't remember even needing to do myself, for any app, scoop-installed or otherwise.

It is `scoop reset xxx@xx.xx.xx`, just this reason that we use `current` junction.

Let's try to implement it, to see if there's something else that needed.

**Comment by rasa** [comment]
> It is `scoop reset xxx@xx.xx.xx`, just this reason that we use `current` junction.

Dang, I knew we could *install* specific versions. I didn't know we could *reset* to a prior version. Tres cool.

**Comment by leumasme** [comment]
https://github.com/winsiderss/systeminformer/issues/1593 relevant issue.
System Informer uses the version-specific application path instead of the `current` path when replacing the windows Task Manager with itself, causing it to break on updates. They have stated that this is a scoop issue.
> The problem is Scoop using symlinks incorrectly. All they need to do is just rename the `current` folder to/from the latest version without using any symlinks:

**Comment by dmex** [comment]
> System Informer uses the version-specific application path instead of the current path

It's not just System Informer... Scoop is currently breaking every application that uses these system functions:

[GetMappedFileName](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getmappedfilenamew)
[GetModuleFileNameEx](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getmodulefilenameexw)
[QueryFullProcessImageName](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-queryfullprocessimagenamew)
[GetProcessImageFileName](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getprocessimagefilenamew)
[Shell_NotifyIcon](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shell_notifyiconw) (calls QueryFullProcessImageName)
[INetFwPolicy2 interface](https://learn.microsoft.com/en-us/windows/win32/api/netfw/nn-netfw-inetfwpolicy2) (uses QueryFullProcessImageName)
... 

@rasa @niheaven This issue needs to be fixed because some of these functions are mandatory for various security reasons. Every application installed by Scoop is only seeing the /apps/appname/xx.xx.xx/ directory instead of /apps/appname/current/.

> https://github.com/winsiderss/systeminformer/issues/1593 relevant issue.

I'll include some information here from the linked issue. The first one is every system call through these /current/ symlinks is currently getting executed twice affecting application performance (see Process Monitor). The second issue is how the Windows kernel is always resolving directory symlinks for security reasons and there are only two solutions (without a kernel driver):
* Create symlinks for files in the /xx.xx.xx/ directory to the /current/ directory (Exes and Dlls can be symlinked without issue and it's only directory symlinks that are resolved).
* Just using the /current/ directory without any symlinks.

> So why don't we simply install into current? Or even ~/scoop/apps/appname? Then we don't even need a persist folder, as current is now "persisted".

Scoop should just use the /current/ directory by default and the /xx.xx.xx/ directory for previous versions.

Rename /appname/Current/ to /appname/xx.xx.xx/
Rename /appname/xx.xx.xx/ to /appname/Current/

If you're using the directory name for versioning then use a file in the /appname/ folder to know what version is /Current/

> It is scoop reset xxx@xx.xx.xx, just this reason that we use current junction.

@niheaven Most users would expect this command to reset the app but it's just resetting the symlink? It should always download xx.xx.xx and just rename the /current/ directory.

Download to /xx.xx.xx/
Rename /Current/ to /xx.xx.xx/
Rename /xx.xx.xx/ to /Current/

**Comment by jshield** [comment]
I personally don't have this issue as my AV software notices the hash of the files have changed and asks me to revalidate firewall/exemptions regardless and the rules are stored a way that I can easily update them via a script as all good AV software should, otherwise you are vulnerable to all sorts of attacks, a physical path is no more secure than a symlink/junction. 

It is a fair point about directory junctions causing issues, particularly with Windows not retaining metadata about the logical path rather than the physical path of the executable image thus:

I would recommend get rid of the junction entirely i.e. set no_junction true by default but keep installing to a separate versioned folder by default.

Windows has this annoying habit of keeping locks on files, and I hate having to close apps just because I am installing a new version of them, particularly when a lot of them are long running background services that have to stay running.

Renaming "reset" to "use" would be clearer with intent.

perhaps the selected install mode (install to new folder and rename to current/install versions side by side/unpack into the same directory) can be selected both as a default and on a per app basis, that way we can chose the most appropriate mode for the app in question.

additionally, being able to update the PATH env var may remove the need for the shims entirely and again should probably be a per app option.

The persist directory is preferred for me as well as it makes backups easier, and smaller, preventing issues where I have to restore the folder and excluding program files (/etc on unix is a godsend for this), so making it a per app option is again probably ideal.

**Comment by pukkandan** [comment]
> I would recommend get rid of the junction entirely i.e. set no_junction true by default but keep installing to a separate versioned folder by default.

That is the total opposite of this request and does not solve any of my problems! If you want such an option, maybe consider opening a separate issue. It is also a terrible default - you may prefer having no fixed path for an app, but most people want it.

> additionally, being able to update the PATH env var may remove the need for the shims entirely and again should probably be a per app option.

Again, this is unrelated to this request, and you may wish to open a new issue requesting it. But imo, this is a bad default as well. Not filling up PATH with junk is one of the good things about scoop.

**Comment by jshield** [comment]
I think you missed my conclusions that this should ultimately be a user selectable behavior per app as applications have different norms with respect to how updates should be handled, and others in this feature discussion are also requesting that we should break existing norms in scoop itself without respecting existing users of the software, as such I believe I have the right to contribute to the discussion of this.

cool that you have done a survey of scoop users that you can show that "most" people want this change in norms, can you share it with the rest of us? I'd like to be able to contribute my own experiences using the software to ensure that the devs.

however I agree that those should probably be spun out into separate items for work tracking purposes if the maintainer of scoop ultimately decides that there is merit.

**Comment by WindyNova** [comment]
This issue is giving me a headache right now

**Comment by osmiumsilver** [comment]
hey guys 🎉, I've made a PowerShell script to workaround this issue, this is my first time contributing to a project, and I’d love to hear your thoughts and feedback! [Link](https://github.com/osmiumsilver/ScoopUpdateWrapper)

**Comment by eugenesvk** [comment]
> Install directly into current and symlink it to the version folder instead.

By the way, if it's possible to avoid the second step of symlinking to the version folder (for example, if you only have one version of the app), then you could have the case of the whole scoop apps directory becoming **more portable** as it'd be easier to just copy&paste it in a file manager (there are other issues with scoop portability like non-relative paths in shims and even persist configs, but those are separate "fixes")
